src/main/java/io/kaitai/struct/Region.java

```diff
@@ -0,0 +1,87 @@
+/**
+ * Copyright 2015-2020 Kaitai Project: MIT license
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+ * LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+ * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package io.kaitai.struct;
+
+/**
+ * Information about positions of parsed value in the streams.
+ *
+ * @since 0.9
+ */
+public class Region {
+  /** Offset from begin of root stream, for which that region was created. */
+  public final long offset;
+  /**
+   * Offset from begin of stream, from which value was parsed. This is relative
+   * position, to get absolute position (relative to root stream) use
+   * {@link #absoluteStart()}. That offset always non-negative.
+   */
+  public final long start;
+  /**
+   * Offset from begin of stream, from which value was parsed. This is relative
+   * position, to get absolute position (relative to root stream) use
+   * {@link #absoluteEnd()}.
+   * <p>
+   * If that offset is negative, then value wasn't parsed yet or exception was
+   * thrown while parsing value.
+   */
+  public long end = -1;
+
+  /**
+   * Creates region that starts at current stream offset and ends at unknown position.
+   *
+   * @param io the stream to get positional information
+   */
+  public Region(KaitaiStream io) {
+    this(io.offset(), io.pos());
+  }
```

Comment:
I personally think this constructor shouldn't exist - I'd prefer to always instantiate it like `new Region(_io.offset(), _io.pos())`. Seeing just `new Region(_io)` in the generated code was a WTF moment for me - why would `Region` need access to the full `KaitaiStream` object? Does it do some I/O operations?

When I see this, it starts to make sense, but it feels wrong.

Reply:
My main motivation to include them was to reduce complexity and possible incompatibilities in the generated code. With such constructor you are free to change how to get position from the stream without touching generated stuff at all (and, for example, replace the return type of the `pos` method in a non-breaking manner). But if you think that this is not enough argument, I'll remove it, just say.

Reply:
@Mingun Fair enough.